src/Catalyst.Core.Lib/Dag/DagNode.cs

```diff
@@ -0,0 +1,371 @@
+        /// </param>
+        /// <returns>
+        ///     A new <see cref="DagNode" /> with the existing and new
+        ///     links.
+        /// </returns>
+        /// <remarks>
+        ///     A <b>DagNode</b> is immutable.
+        /// </remarks>
+        public IDagNode AddLinks(IEnumerable<IMerkleLink> links)
+        {
+            var all = Links.Union(links);
+            return new DagNode(DataBytes, all, _hashAlgorithm);
+        }
+
+        /// <summary>
+        ///     Removes a link.
+        /// </summary>
+        /// <param name="link">
+        ///     The <see cref="IMerkleLink" /> to remove.
+        /// </param>
+        /// <returns>
+        ///     A new <see cref="DagNode" /> with the <paramref name="link" />
+        ///     removed.
+        /// </returns>
+        /// <remarks>
+        ///     A <b>DagNode</b> is immutable.
+        ///     <para>
+        ///         No exception is raised if the <paramref name="link" /> does
+        ///         not exist.
+        ///     </para>
+        /// </remarks>
+        public DagNode RemoveLink(IMerkleLink link) { return RemoveLinks(new[] {link}); }
+
+        /// <summary>
+        ///     Remove a sequence of links.
+        /// </summary>
+        /// <param name="links">
+        ///     The sequence of <see cref="IMerkleLink" /> to remove.
+        /// </param>
+        /// <returns>
+        ///     A new <see cref="DagNode" /> with the <paramref name="links" />
+        ///     removed.
+        /// </returns>
+        /// <remarks>
+        ///     A <b>DagNode</b> is immutable.
+        ///     <para>
+        ///         No exception is raised if any of the <paramref name="links" /> do
+        ///         not exist.
+        ///     </para>
+        /// </remarks>
+        public DagNode RemoveLinks(IEnumerable<IMerkleLink> links)
+        {
+            var ignore = links.ToLookup(link => link.Id);
+            var some = Links.Where(link => !ignore.Contains(link.Id));
+            return new DagNode(DataBytes, some, _hashAlgorithm);
+        }
+
+        /// <summary>
+        ///     Writes the binary representation of the node to the specified <see cref="Stream" />.
+        /// </summary>
+        /// <param name="stream">
+        ///     The <see cref="Stream" /> to write to.
+        /// </param>
+        public void Write(Stream stream)
+        {
+            using (var cos = new CodedOutputStream(stream, true))
+            {
+                Write(cos);
+            }
+        }
+
+        /// <summary>
+        ///     Writes the binary representation of the node to the specified <see cref="CodedOutputStream" />.
+        /// </summary>
+        /// <param name="stream">
+        ///     The <see cref="CodedOutputStream" /> to write to.
+        /// </param>
+        public void Write(CodedOutputStream stream)
+        {
+            if (stream == null)
+                throw new ArgumentNullException(nameof(stream));
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The parameter name '' is not declared in the argument list.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)